src/field/scrcmd_mystery_gift.c

```diff
@@ -0,0 +1,684 @@
+
+static const struct ScriptMysteryGiftFuncs sScriptMysteryGiftActionTable[MG_TAG_MAX - 1] = {
+    { MGCheck_PartySpace, MGGive_Mon, MGMessageSuccess_GiveMon, MGMessageFailure_GiveMon },
+    { MGCheck_PartySpace, MGGive_Egg, MGMessageSuccess_Egg, MGMessageFailure_GiveMon },
+    { MGCheck_Item, MGGive_Item, MGMessageSuccess_Item, MGMessageFailure_Item },
+    { MGCheck_BattleRules, MGGive_BattleRules, MGMessageSuccess_BattleRules, MGMessageFailure_BattleRules },
+    { MGCheck_Decoration, MGGive_Decoration, MGMessageSuccess_Decoration, MGMessageFailure_Decoration },
+    { MGCheck_MonDeco, MGGive_MonDeco, MGMessageSuccess_MonDeco, MGMessageFailure_MonDeco },
+    { MGCheck_PartySpace, MGGive_ManaphyEgg, MGMessageSuccess_ManaphyEgg, MGMessageFailure_GiveMon },
+    { MGCheck_MemberCard, MGGive_MemberCard, MGMessageSuccess_MemberCard, MGMessageFailure_MemberCard },
+    { MGCheck_OaksLetter, MGGive_OaksLetter, MGMessageSuccess_OaksLetter, MGMessageFailure_OaksLetter },
+    { MGCheck_AzureFlute, MGGive_AzureFlute, MGMessageSuccess_AzureFlute, MGMessageFailure_AzureFlute },
+    { MGCheck_PoketchApp, MGGive_PoketchApp, MGMessageSuccess_PoketchApp, MGMessageFailure_PoketchApp },
+    { MGCheck_SecretKey, MGGive_SecretKey, MGMessageSuccess_SecretKey, MGMessageFailure_SecretKey },
+    { MGCheck_PartySpace, MGGive_Mon, MGMessageSuccess_GiveMon, MGMessageFailure_GiveMon },
+    { MGCheck_PokewalkerCourse, MGGive_PokewalkerCourse, MGMessageSuccess_PokewalkerCourse, MGMessageFailure_PokewalkerCourse },
+    { MGCheck_MemorialPhoto, MGGive_MemorialPhoto, MGMessageSuccess_MemorialPhoto, MGMessageFailure_MemorialPhoto },
+};
+
+void FieldSystem_InitGetMysteryGiftGmmState(struct GetMysteryGiftGmmState* state, FieldSystem* fieldSys, MessageFormat* msgFormat, MysteryGiftData* mgData) {
+    state->fieldSys = fieldSys;
+    state->msgFormat = msgFormat;
+    state->mgData = mgData;
+}
+
+int FieldSystem_GetTagOfNextMG(FieldSystem* fieldSys) {
+    (void)fieldSys;
+    return SaveMGDataPtr_GetTagByIndex(SaveMGDataPtr_GetFirstGiftIndex());
+}
+
+MysteryGiftData* FieldSystem_GetDataOfNextMG(FieldSystem* fieldSys) {
+    (void)fieldSys;
+    return SaveMGDataPtr_GetDataByIndex(SaveMGDataPtr_GetFirstGiftIndex());
+}
+
+void FieldSystem_SetQueuedMGReceived(FieldSystem* fieldSys) {
+    (void)fieldSys;
+    SaveMGDataPtr_SetReceivedByIndex(SaveMGDataPtr_GetFirstGiftIndex());
+}
+
+BOOL ScrCmd_MysteryGift(ScriptContext* ctx) {
+    switch (ScriptReadHalfword(ctx)) {
+    case SCR_MG_BEGIN:
+        SaveMGDataPtr_Begin(ctx->fieldSystem->saveData, HEAP_ID_32);
+        break;
+    case SCR_MG_END:
+        SaveMGDataPtr_End(ctx->fieldSystem->saveData, FALSE);
+        break;
+    case SCR_MG_END2:
+        SaveMGDataPtr_End(ctx->fieldSystem->saveData, TRUE);
+        break;
+    case SCR_MG_HAS_GIFT: {
+        u16* ptr = ScriptGetVarPointer(ctx);
+        if (FieldSystem_GetTagOfNextMG(ctx->fieldSystem) != MG_TAG_INVALID) {
```

Comment:
does `*ptr = FieldSystem_GetTagOfNextMG(ctx->fieldSystem) != MG_TAG_INVALID` work?

Reply:
no. no it does not.